File: java/src/array/SinglyTailLinkedQueue.java
package array;

public class SinglyTailLinkedQueue<E> {

    private Node head;

    private Node tail;

    private int size;

    private class Node {
        private E key;

        private Node next;

        public Node(E key) {
            this.key = key;
        }
    }

    public String toString() {
        Node tmp = head;

        StringBuffer sb = new StringBuffer();
        sb.append("[");
        for(int i = 0 ; i < size ; i++) {
            sb.append(tmp.key);
            sb.append(",");
            tmp = tmp.next;
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append("]");

        return sb.toString();
    }

    public void enQueue(E key) {
        Node node = new Node(key);

        if (head == null)
            head = node;

        if (tail != null)
            tail.next = node;

        tail = node;
        size++;
    }

    public E deQueue() {
        if (head == null)
            return null;

        Node del = head;
        head = del.next;

        E returnValue = del.key;
        del = null;
        size--;

        return returnValue;
    }

    public boolean empty() {
        return head == null;
    }
}

File: python/array/singlyTailLinkedQueue.py
class SinglyTailLinkedQueue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    class Node:
        def __init__(self, data):
            self.data = data
            self.next = None

    def toString(self):
        tmp = self.head
        result = []
        for i in range(self.size):
            result.append(str(tmp.data))
            tmp = tmp.next
        return ','.join(result)


    def enqueue(self, input):
        node = self.Node(input)
        if self.tail == None:
            self.head = node
        else:
            self.tail.next = node
        self.tail = node
        self.size += 1

    def dequeue(self):
        if self.head == None:
            return None
        node = self.head
        result = node.data
        self.head = node.next
        node = None
        self.size -= 1
        return result

    def empty(self):
        return self.head == None
